File: ios/Sources/NativeAgentPlugin/MemoryProviderImpl.swift
import Foundation

#if canImport(CapacitorLanceDB) || canImport(CapacitorLancedb) || canImport(LanceDBPlugin)
public final class MemoryProviderImpl: MemoryProvider {
    public static func makeIfAvailable() -> MemoryProvider? {
        guard LanceDBBridge.shared.getOrCreateHandle() != nil else {
            return nil
        }
        return MemoryProviderImpl()
    }

    public func store(key: String, text: String, metadataJson: String?) -> String {
        guard let handle = LanceDBBridge.shared.getOrCreateHandle() else {
            return errorJson("Memory provider not configured")
        }

        let resolvedKey = key.isEmpty
            ? "mem-\(Int(Date().timeIntervalSince1970 * 1000))-\(UUID().uuidString.prefix(8))"
            : key
        let embedding = localHashEmbed(text, dim: Int(LanceDBBridge.embeddingDim))

        do {
            try LanceDBBridge.awaitResult {
                try await handle.store(
                    key: resolvedKey,
                    agentId: Self.defaultAgentId,
                    text: text,
                    embedding: embedding,
                    metadata: metadataJson
                )
            }
            return jsonString([
                "success": true,
                "key": resolvedKey,
            ])
        } catch {
            return errorJson(error.localizedDescription)
        }
    }

    public func recall(query: String, limit: UInt32) -> String {
        guard let handle = LanceDBBridge.shared.getOrCreateHandle() else {
            return errorJson("Memory provider not configured")
        }

        let embedding = localHashEmbed(query, dim: Int(LanceDBBridge.embeddingDim))

        do {
            let results = try LanceDBBridge.awaitResult {
                try await handle.search(
                    queryVector: embedding,
                    limit: limit,
                    filter: "agent_id = '\(Self.defaultAgentId)'"
                )
            }
            return jsonString(results.map { result in
                var object: [String: Any] = [
                    "key": result.key,
                    "text": result.text,
                    "score": result.score,
                ]
                if let metadata = result.metadata {
                    object["metadata"] = metadata
                }
                return object
            })
        } catch {
            return errorJson(error.localizedDescription)
        }
    }

    public func forget(key: String) -> String {
        guard let handle = LanceDBBridge.shared.getOrCreateHandle() else {
            return errorJson("Memory provider not configured")
        }
        guard !key.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return errorJson("Provide query or key.")
        }

        do {
            try LanceDBBridge.awaitResult {
                try await handle.delete(key: key)
            }
            return jsonString([
                "success": true,
                "key": key,
            ])
        } catch {
            return errorJson(error.localizedDescription)
        }
    }

    public func search(query: String, maxResults: UInt32) -> String {
        guard let handle = LanceDBBridge.shared.getOrCreateHandle() else {
            return errorJson("Memory provider not configured")
        }

        let embedding = localHashEmbed(query, dim: Int(LanceDBBridge.embeddingDim))

        do {
            let results = try LanceDBBridge.awaitResult {
                try await handle.search(
                    queryVector: embedding,
                    limit: maxResults,
                    filter: "agent_id = '\(Self.defaultAgentId)'"
                )
            }
            return jsonString(results.map { result in
                var object: [String: Any] = [
                    "key": result.key,
                    "text": result.text,
                    "score": result.score,
                ]
                if let metadata = result.metadata {
                    object["metadata"] = metadata
                }
                return object
            })
        } catch {
            return errorJson(error.localizedDescription)
        }
    }

    public func list(prefix: String?, limit: UInt32?) -> String {
        guard let handle = LanceDBBridge.shared.getOrCreateHandle() else {
            return errorJson("Memory provider not configured")
        }

        do {
            let keys = try LanceDBBridge.awaitResult {
                try await handle.list(prefix: prefix, limit: limit)
            }
            return jsonString(keys)
        } catch {
            return errorJson(error.localizedDescription)
        }
    }

    private static let defaultAgentId = "main"
    private static let whitespace = CharacterSet.whitespacesAndNewlines
    private static let posixLocale = Locale(identifier: "en_US_POSIX")
    private static let fnvOffset: UInt32 = 0x811c9dc5
    private static let fnvPrime: UInt32 = 0x01000193
    private static let goldenRatio: UInt32 = 2654435761
    private static let mixConstant: UInt32 = 0x45d9f3b
    private static let uintMaxDouble = Double(UInt32.max)

    private func jsonString(_ value: Any) -> String {
        guard JSONSerialization.isValidJSONObject(value),
              let data = try? JSONSerialization.data(withJSONObject: value),
              let string = String(data: data, encoding: .utf8) else {
            return "{\"error\":\"Failed to encode JSON\"}"
        }
        return string
    }

    private func errorJson(_ message: String) -> String {
        let escaped = message
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "\"", with: "\\\"")
        return "{\"error\":\"\(escaped)\"}"
    }

    private func fnv1a(_ text: String) -> UInt32 {
        var hash = Self.fnvOffset
        for scalar in text.unicodeScalars {
            hash ^= UInt32(scalar.value)
            hash = hash &* Self.fnvPrime
        }
        return hash
    }

    private func seededRandom(seed: UInt32, dim: Int) -> Float {
        var h = seed ^ (UInt32(dim) &* Self.goldenRatio)
        h = ((h >> 16) ^ h) &* Self.mixConstant
        h = ((h >> 16) ^ h) &* Self.mixConstant
        h = (h >> 16) ^ h
        return Float((Double(h) / Self.uintMaxDouble) * 2.0 - 1.0)
    }

    private func localHashEmbed(_ text: String, dim: Int) -> [Float] {
        let tokens = text
            .lowercased(with: Self.posixLocale)
            .components(separatedBy: Self.whitespace)
            .filter { !$0.isEmpty }
        var vector = Array(repeating: 0.0, count: dim)

        for token in tokens {
            let hash = fnv1a(token)
            for index in 0..<dim {
                vector[index] += Double(seededRandom(seed: hash, dim: index))
            }
        }

        let norm = sqrt(vector.reduce(0.0) { $0 + ($1 * $1) })
        let divisor = norm > 0 ? norm : 1
        return vector.map { Float($0 / divisor) }
    }
}
#else
public final class MemoryProviderImpl {
    public static func makeIfAvailable() -> MemoryProvider? {
        nil
    }
}
#endif

File: android/src/main/java/com/t6x/plugins/nativeagent/MemoryProviderImpl.kt
package com.t6x.plugins.nativeagent

import android.content.Context
import java.util.Locale
import java.util.UUID
import java.util.concurrent.Executors
import kotlinx.coroutines.runBlocking
import org.json.JSONArray
import org.json.JSONObject
import uniffi.lancedb_ffi.HybridSearchResult
import uniffi.lancedb_ffi.SearchResult
import uniffi.native_agent_ffi.MemoryProvider

class MemoryProviderImpl(private val context: Context) : MemoryProvider {
    fun isAvailable(): Boolean = LanceDBBridge.getOrCreateHandle(context) != null

    // UniFFI methods on LanceDbHandle are suspend → require runBlocking. The
    // MemoryProvider callback is invoked by Rust on a JNA-managed native
    // thread; running coroutines (and uniffiRustCallAsync's own JNA callbacks)
    // on that thread races with JNA's thread-detach logic, producing
    // SIGABRT "attempting to detach while still running code" on Android ART.
    // Hop to a dedicated JVM-owned worker thread so the JNA callback thread
    // only blocks on Future.get(), never executes coroutine continuations.
    private fun <T> onWorker(block: () -> T): T = WORKER.submit(block).get()

    override fun store(key: String, text: String, metadataJson: String?): String {
        val handle = LanceDBBridge.getOrCreateHandle(context) ?: return unavailableJson()
        val resolvedKey = key.ifBlank {
            "mem-${System.currentTimeMillis()}-${UUID.randomUUID().toString().take(8)}"
        }
        val embedding = localHashEmbed(text, LanceDBBridge.EMBEDDING_DIM)
        return runCatching {
            onWorker {
                runBlocking {
                    handle.store(resolvedKey, DEFAULT_AGENT_ID, text, embedding, metadataJson)
                }
            }
            JSONObject()
                .put("success", true)
                .put("key", resolvedKey)
                .toString()
        }.getOrElse { errorJson(it) }
    }

    override fun recall(query: String, limit: UInt): String {
        val handle = LanceDBBridge.getOrCreateHandle(context) ?: return unavailableJson()
        val embedding = localHashEmbed(query, LanceDBBridge.EMBEDDING_DIM)
        return runCatching {
            val results = onWorker {
                runBlocking {
                    handle.hybridSearch(
                        embedding,
                        query,
                        limit,
                        "agent_id = '$DEFAULT_AGENT_ID'",
                        null,
                        null,
                    )
                }
            }
            hybridResultsJson(results)
        }.getOrElse { errorJson(it) }
    }

    override fun forget(key: String): String {
        val handle = LanceDBBridge.getOrCreateHandle(context) ?: return unavailableJson()
        if (key.isBlank()) {
            return JSONObject().put("error", "Provide query or key.").toString()
        }
        return runCatching {
            onWorker {
                runBlocking {
                    handle.delete(key)
                }
            }
            JSONObject()
                .put("success", true)
                .put("key", key)
                .toString()
        }.getOrElse { errorJson(it) }
    }

    override fun search(query: String, maxResults: UInt): String {
        val handle = LanceDBBridge.getOrCreateHandle(context) ?: return unavailableJson()
        val embedding = localHashEmbed(query, LanceDBBridge.EMBEDDING_DIM)
        return runCatching {
            val results = onWorker {
                runBlocking {
                    handle.search(embedding, maxResults, "agent_id = '$DEFAULT_AGENT_ID'")
                }
            }
            searchResultsJson(results)
        }.getOrElse { errorJson(it) }
    }

    override fun list(prefix: String?, limit: UInt?): String {
        val handle = LanceDBBridge.getOrCreateHandle(context) ?: return unavailableJson()
        return runCatching {
            val keys = onWorker {
                runBlocking {
                    handle.list(prefix, limit)
                }
            }
            JSONArray(keys).toString()
        }.getOrElse { errorJson(it) }
    }

    private fun hybridResultsJson(results: List<HybridSearchResult>): String {
        val array = JSONArray()
        results.forEach { result ->
            array.put(
                JSONObject()
                    .put("key", result.key)
                    .put("text", result.text)
                    .put("vectorRank", result.vectorRank.toInt())
                    .put("textRank", result.textRank.toInt())
                    .put("rrfScore", result.rrfScore)
                    .apply {
                        result.metadata?.let { put("metadata", it) }
                    }
            )
        }
        return array.toString()
    }

    private fun searchResultsJson(results: List<SearchResult>): String {
        val array = JSONArray()
        results.forEach { result ->
            array.put(
                JSONObject()
                    .put("key", result.key)
                    .put("text", result.text)
                    .put("score", result.score)
                    .apply {
                        result.metadata?.let { put("metadata", it) }
                    }
            )
        }
        return array.toString()
    }

    private fun unavailableJson(): String {
        return JSONObject().put("error", "Memory provider not configured").toString()
    }

    private fun errorJson(error: Throwable): String {
        return JSONObject()
            .put("error", error.message ?: error::class.java.simpleName)
            .toString()
    }

    private companion object {
        private const val DEFAULT_AGENT_ID = "main"
        private val WORKER = Executors.newSingleThreadExecutor { r ->
            Thread(r, "memory-provider-worker").apply { isDaemon = true }
        }
        private val WHITESPACE = Regex("\\s+")
        private val UINT_MAX_DOUBLE = 0xffffffffu.toDouble()
        private const val FNV_OFFSET = 0x811c9dc5u
        private const val FNV_PRIME = 0x01000193u
        private const val GOLDEN_RATIO = 2654435761u
        private const val MIX = 0x45d9f3bu

        private fun fnv1a(text: String): UInt {
            var hash = FNV_OFFSET
            text.forEach { char ->
                hash = hash xor char.code.toUInt()
                hash *= FNV_PRIME
            }
            return hash
        }

        private fun seededRandom(seed: UInt, dim: Int): Float {
            var h = seed xor (dim.toUInt() * GOLDEN_RATIO)
            h = (((h shr 16) xor h) * MIX)
            h = (((h shr 16) xor h) * MIX)
            h = (h shr 16) xor h
            return (((h.toDouble() / UINT_MAX_DOUBLE) * 2.0) - 1.0).toFloat()
        }

        private fun localHashEmbed(text: String, dim: Int): List<Float> {
            val tokens = text.lowercase(Locale.ROOT).split(WHITESPACE).filter { it.isNotBlank() }
            val vector = DoubleArray(dim)
            tokens.forEach { token ->
                val hash = fnv1a(token)
                for (index in 0 until dim) {
                    vector[index] += seededRandom(hash, index).toDouble()
                }
            }

            var norm = 0.0
            for (value in vector) {
                norm += value * value
            }
            norm = kotlin.math.sqrt(norm).takeIf { it > 0.0 } ?: 1.0

            return List(dim) { index -> (vector[index] / norm).toFloat() }
        }
    }
}
